src/xrpld/app/tx/detail/MPTokenAuthorize.cpp

```diff
@@ -0,0 +1,251 @@
+//------------------------------------------------------------------------------
+/*
+  This file is part of rippled: https://github.com/ripple/rippled
+  Copyright (c) 2023 Ripple Labs Inc.
+
+  Permission to use, copy, modify, and/or distribute this software for any
+  purpose  with  or without fee is hereby granted, provided that the above
+  copyright notice and this permission notice appear in all copies.
+
+  THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+  WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+  MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+  ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+  WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+  ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/MPTokenAuthorize.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/st.h>
+
+namespace ripple {
+
+NotTEC
+MPTokenAuthorize::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureMPTokensV1))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfMPTokenAuthorizeMask)
+        return temINVALID_FLAG;
+
+    if (ctx.tx[sfAccount] == ctx.tx[~sfMPTokenHolder])
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+MPTokenAuthorize::preclaim(PreclaimContext const& ctx)
+{
+    auto const accountID = ctx.tx[sfAccount];
+    auto const holderID = ctx.tx[~sfMPTokenHolder];
+
+    if (holderID && !(ctx.view.exists(keylet::account(*holderID))))
+        return tecNO_DST;
+
+    // if non-issuer account submits this tx, then they are trying either:
+    // 1. Unauthorize/delete MPToken
+    // 2. Use/create MPToken
+    //
+    // Note: `accountID` is holder's account
+    //       `holderID` is NOT used
+    if (!holderID)
+    {
+        std::shared_ptr<SLE const> sleMpt = ctx.view.read(
+            keylet::mptoken(ctx.tx[sfMPTokenIssuanceID], accountID));
+
+        // There is an edge case where holder deletes MPT after issuance has
+        // already been destroyed. So we must check for unauthorize before
+        // fetching the MPTIssuance object(since it doesn't exist)
+
+        // if holder wants to delete/unauthorize a mpt
+        if (ctx.tx.getFlags() & tfMPTUnauthorize)
+        {
+            if (!sleMpt)
+                return tecOBJECT_NOT_FOUND;
+
+            if ((*sleMpt)[sfMPTAmount] != 0)
+                return tecHAS_OBLIGATIONS;
```

Comment:
If the MPT has a balance, then the issuance should exist. If it does not, that would be Bad.
```suggestion
            if ((*sleMpt)[sfMPTAmount] != 0) {
                auto const sleMptIssuance =
            ctx.view.read(keylet::mptIssuance(ctx.tx[sfMPTokenIssuanceID]));
                assert(sleMptIssuance);
                if (!sleMptIssuance)
                    return tefINTERNAL;
                return tecHAS_OBLIGATIONS;
            }
```

Reply:
https://github.com/gregtatcam/rippled/pull/44